### ISSUE TITLE
thrust::find doesn't handle asymmetric equality operator

### ISSUE THREAD

**Issue by dkolsen-pgi** [thrust/thrust#1229]
`thrust::find` fails to compile when the input iterator's value type is different from the type of the value being searched for and neither type is convertible to the other, even when there is an overloaded `operator==` that takes objects of those two different types.

Given the following CUDA program:
```
#include <thrust/host_vector.h>
#include <thrust/device_vector.h>
#include <thrust/find.h>
#include <cassert>

class Weird {
  int value;
 public:
  friend __host__ __device__ bool operator==(int x, Weird y) {
    return x == y.value;
  }
  __host__ __device__ Weird(int val, int) : value(val) { }
};

int main() {
  thrust::host_vector<int> v;
  for (int i = 0; i < 10000; ++i) {
    v.push_back(i);
  }
  thrust::device_vector<int> dv(v);
  auto result = thrust::find(dv.begin(), dv.end(), Weird(333, 0));
  assert(*result == 333);
}
```
When compiled with NVCC 11.0 and the latest Thrust source from the main branch, this fails with:
```
/proj/cuda/thrust/master/thrust/detail/functional/operators/operator_adaptors.h(108): error: function "thrust::equal_to<T>::operator() [with T=int]" cannot be ca
lled with the given argument list
            argument types are: (int, Weird)
            object type is: thrust::equal_to<int>
```
(followed by a typical template instantiation novel).

This is a recent regression.  It broke somewhere between July 3 and July 6.  I have not looked to see which commit might have caused the regression.  When compiled with the Thrust in CUDA 11.0, this test case compiles successfully.

**Comment by allisonvacanti** [comment]
It is likely related to 0b48df1fcef6de19bf611a0ffc236b1c8871669f.

It looks like adding a `equal_to<void>` specialization (like in thrust/functional.h in b5284512e83bf91eac9010ef370ce27988ae091f) should fix this.

**Comment by dkolsen-pgi** [comment]
I just realized that the regression is in the 1.9.10-1 branch, not just in main.  This is a regression in our soon-to-be-code-frozen release.  Please get the fix into 1.9.10-1 ASAP.

**Comment by allisonvacanti** [comment]
If it's what I think it is, we can revert the change in the staging branch ASAP. We should be able to just revert the commit that was likely to introduce this.

If you get a chance to test this out tonight let me know, but otherwise I'll test it out and integrate it tomorrow.

In your thrust git repo, checkout the staging branch and revert f483906c28e6f7fe9b10eed7574a5ed027fe8690.
```
git checkout staging/1.9.10-1
git revert f483906
```

**Comment by allisonvacanti** [comment]
Confirmed that reverting f483906 fixes the regression. Fixed in perforce CL 28805582 and `staging/1.9.10-1` commit c3c368a69adc580eb35f45f78760be2612441899.

Let's leave this issue open until we properly address this in `main`.

**Comment by bjude** [comment]
this issue is also present with `thrust::equal`, which delegates down to `thrust::equal_to<T>`, not sure if it's resolved by the commits Allison referenced

**Comment by allisonvacanti** [comment]
@bjude All issues from using heterogeneous types in the placeholder code will be fixed by that patch. If you have a small reproduction of the case you encountered, please share it and I'll verify/add a regression test for it.

**Comment by allisonvacanti** [comment]
Fixed in #1237.
